Source/WebCore/dom/Document.cpp

```diff
@@ -977,6 +977,18 @@ Ref<SecurityOrigin> Document::protectedTopOrigin() const
     return topOrigin();
 }
 
+SecurityOrigin& Document::topOrigin() const
+{
+    RefPtr frame = this->frame();
+    if (RefPtr localMainFrame = frame ? dynamicDowncast<LocalFrame>(frame->mainFrame()) : nullptr) {
+        if (RefPtr mainFrameDocument = localMainFrame->document())
+            return mainFrameDocument->securityOrigin();
+    }
+    if (RefPtr page = this->page())
+        return page->mainFrameOrigin();
```

Comment:
In existing implementation, we run check on security policy (`SecurityContext::securityOrigin()`) for `Document::topOrigin()`, I don't see the security check in this patch. Do we run it when setting main frame URL for the subframe process?

Reply:
What checks are done in SecurityContext::securityOrigin?  We either don't need to do them in a remote frame process or we haven't implemented them for site isolation yet.

Reply:
SecurityContext::securityOrigin is decided by securityOriginPolicy, and securityOriginPolicy can be set via SecurityContext::setSecurityOriginPolicy. It seems in most cases it will be SecurityOrigin of the main document, but can be opaque (e.g. if the iframe is sandboxed).

Reply:
I think that goes in the category of things that haven't been implemented in site isolation yet, but this PR is an improvement still.

Reply:
ok let's add a FIXME about adding check for SecurityOriginPolicy of top document later.

Reply:
Great idea!